test/RefactoredTestSetup.ts

```diff
@@ -0,0 +1,360 @@
+      await utilities.reloadAndEnableExtensions(); // This is necesssary in order to update JAVA home path
+    }
+  }
+
+  public async tearDown(): Promise<void> {
+    await this.checkForUncaughtErrors();
+    try {
+      await utilities.deleteScratchOrg(this.scratchOrgAliasName);
+      await this.deleteScratchOrgInfo();
+      this.projectFolderPath = undefined;
+      this.scratchOrgAliasName = undefined;
+      this.scratchOrgId = undefined;
+    } catch (error) {
+      utilities.log(
+        `Deleting scratch org (or info) failed with Error: ${(error as Error).message}`
+      );
+    }
+  }
+
+  public async setUpTestingWorkspace(projectConfig: ProjectConfig) {
+    utilities.log(`${this.testSuiteSuffixName} - Starting setUpTestingWorkspace()...`);
+    let projectName;
+    switch (projectConfig.projectShape) {
+      case ProjectShapeOption.NEW:
+        if (!fs.existsSync(this.tempFolderPath)) {
+          utilities.createFolder(this.tempFolderPath);
+        }
+        await utilities.generateSfProject(this.tempProjectName, this.tempFolderPath); // generate a sf project for 'new'
+        this.projectFolderPath = path.join(this.tempFolderPath, this.tempProjectName);
+        break;
+
+      case ProjectShapeOption.NAMED:
+        if (projectConfig.githubRepoUrl) {
+          // verify if folder matches the github repo url
+          const repoExists = await utilities.gitRepoExists(projectConfig.githubRepoUrl);
+          if (!repoExists) {
+            this.throwError(`Repository does not exist or is inaccessible: ${projectConfig.githubRepoUrl}`);
+          }
+          const repoName = utilities.getRepoNameFromUrl(projectConfig.githubRepoUrl);
+          projectName = repoName;
+          if (projectConfig.folderPath) {
+            const localProjName = utilities.getFolderName(projectConfig.folderPath);
+            if (localProjName != repoName) {
+              this.throwError(`the local project ${projectConfig.githubRepoUrl} does not match the required Github repo ${projectConfig.githubRepoUrl}`);
+            } else {
+              // if it is a match, use the local folder directly. Local dev use only.
+              this.projectFolderPath = projectConfig.folderPath;
+            }
+          } else {
+            // clone the project from Github URL directly
+            this.projectFolderPath = path.join(this.tempFolderPath, repoName!);
+            await utilities.gitClone(projectConfig.githubRepoUrl, this.projectFolderPath);
+          }
+        } else {
+          // missing info, throw an error
+          this.throwError(`githubRepoUrl is required for named project shape`);
+        }
+        break;
+
+      case ProjectShapeOption.ANY:
+        // ANY: workspace is designated to open when wdio is initialized
+        if (projectConfig.folderPath) {
+          this.projectFolderPath = projectConfig.folderPath;
+          projectName = utilities.getFolderName(projectConfig.folderPath);
```

Comment:
1. On GHA, we can always do git clone before wdio is started
2. On local, we can also always clone before we run test locally

We can support do git clone when wdio is run but I do not see a necessity for it, and also when running on wdio, we try to avoid dummy workspace opening. So I prefer to only support starting with the target project once wdio is run.

Reply:
I meant, what if you take the project url as an additional input in the workflow dispatch for the "any" case?  The project from that input url would be cloned before WDIO is started.

Reply:
Do you mean something I did [here](https://github.com/forcedotcom/salesforcedx-vscode/pull/5730/files#diff-9ae54a92531817cc2244971629d759a5ceb94fb6992b665fcfd57fdbc46af0ad)?

Reply:
Yes! Use the `githubRepoUrl` input as the “any” repo that you want to clone for the current E2E test. 🙂